rakelib/docs.rake

```diff
@@ -196,25 +225,71 @@ begin
       $stdout.puts "Generated bolt-defaults.yaml reference at:\n\t#{filepath}"
     end
 
+    desc "Generate markdown docs for Bolt's data types"
+    task type_reference: :generate_strings do
+      filepath = File.expand_path('../documentation/bolt_types_reference.md', __dir__)
+      template = File.expand_path('../documentation/templates/bolt_types_reference.md.erb', __dir__)
+
+      @types = @puppet_strings['data_types'].map do |data|
+        functions = []
+
+        # Turn types that include TypeReference into just the name of the
+        # type.
+        data = Bolt::Util.walk_vals(data) do |val|
+          if val.is_a?(String)
+            val.gsub(/TypeReference\['(\w*)'\]/, '\1')
+          else
+            val
+          end
+        end
+
+        # Add attributes, which are listed under the type's params
+        data.dig('docstring', 'tags').each do |param|
+          functions << {
+            'name' => param['name'],
+            'desc' => param['text'].tr("\n", ' '),
+            'type' => param['types'].first
```

Comment:
It'd be nice to strip the `[1]` from Strings here if we can, including in types like `Variant[String[1], Type[Resource]]`. I think the best way to achieve that is with a regex: `.gsub(/String\[1\]/)`. I think we should only do that for `String[1]` though, if we have a type that requires a minimum character length of more than that that seems important to call out here.